File: data_structures_and_algorithm/linked_list/singly_linked_list/insert_end.c
// insert_end.c
// This code defines a simple singly linked list in C and provides a function to insert a new node at the end of the list.
// It also includes a function to print the linked list for demonstration purposes. 

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* insert_at_end(struct Node* head, int data) {
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;

    if (head == NULL)
        return new_node;

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;

    temp->next = new_node;
    return head;
}

void print_list(struct Node* head) {
    while (head != NULL) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main() {
    struct Node* head = NULL;
    head = insert_at_end(head, 10);
    head = insert_at_end(head, 20);
    head = insert_at_end(head, 30);

    printf("Linked List after insertions:\n");
    print_list(head);
    return 0;
}

File: Data Structures and Algorithm/Linked_List/Singly_Linked_List/insert_end.py
# insert_end.py
# This code implements a function to insert a node at the end of a singly linked list.
# It defines a Node class and provides functions to insert a node and print the list.
# The code includes an example usage of the insert function.
# The insert function traverses the list to find the end and appends the new node,
# maintaining the integrity of the singly linked list structure.
# The print function displays the list in a readable format.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insert_at_end(head, data):
    new_node = Node(data)
    if head is None:
        return new_node
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    return head

def print_list(head):
    while head:
        print(head.data, end=" -> ")
        head = head.next
    print("None")

if __name__ == "__main__":
    head = None
    for value in [10, 20, 30]:
        head = insert_at_end(head, value)
    print("Linked List after insertions:")
    print_list(head)
